runtime-extension-model/src/main/java/org/mule/runtime/core/internal/extension/ForEachCollectionTypeResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.extension;
+
+import static org.mule.runtime.api.metadata.resolving.FailureCode.UNKNOWN;
+import static org.mule.runtime.core.api.extension.provider.MuleExtensionModelProvider.ANY_TYPE;
+
+import static java.lang.String.format;
+
+import org.mule.metadata.api.model.ArrayType;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.message.api.el.TypeBindings;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.metadata.ExpressionLanguageMetadataService;
+import org.mule.runtime.api.metadata.MetadataContext;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.InputTypeResolver;
+import org.mule.runtime.privileged.metadata.InternalMetadataContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link InputTypeResolver} implementation used for the "collection" parameter of the ForEach and ParallelForEach scopes.
+ *
+ * @since 4.8.0
+ */
+public class ForEachCollectionTypeResolver implements InputTypeResolver<String> {
+
+  @Override
+  public String getCategoryName() {
+    return "FOREACH";
+  }
+
+  @Override
+  public String getResolverName() {
+    return "FOREACH_INPUT_COLLECTION";
+  }
+
+  @Override
+  public MetadataType getInputMetadata(MetadataContext context, String key)
+      throws MetadataResolvingException, ConnectionException {
+    if (context instanceof InternalMetadataContext) {
+      InternalMetadataContext internalContext = (InternalMetadataContext) context;
+      if (internalContext.getExpressionLanguageMetadataService().isPresent() && internalContext.getTypeBindings().isPresent()) {
+        String expression = sanitizeExpression(key);
+        MetadataType resolvedType = resolveExpressionType(internalContext.getTypeBindings().get(),
+                                                          internalContext.getExpressionLanguageMetadataService().get(),
+                                                          expression);
+        if (!(resolvedType instanceof ArrayType)) {
+          throw new IllegalArgumentException(format("Expression `%s` does not resolve to a collection", expression));
+        }
+        return ((ArrayType) resolvedType).getType();
+      }
+    }
+
+    // Fallback to any
+    return ANY_TYPE;
+  }
+
+  private String sanitizeExpression(String expression) {
```

Comment:
isn't this exact logic already present someplace else. This definitively rings a bell

Reply:
Yes, but I couldn't find any implementation that has been apified that I could reuse.
One is in [core](https://github.com/mulesoft/mule/blob/master/core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageUtils.java#L65).
Another one in [mule-framework](https://github.com/mulesoft/mule-framework/blob/master/mule-framework/src/main/java/org/mule/framework/internal/tooling/type/propagation/cfg/TypePropagationCFGVisitor.java#L567).
Maybe we can apify this on [MuleExpressionLanguage](https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/el/MuleExpressionLanguage.java#L20), what do you think?

Reply:
have mixed feelings about the expression language being in charge of sanitizing its own expressions, although I see its merits... @elrodro83 what do you think?

As for the impl, I would apify the core one. THe framework one doesn't seem to be doing the exact same thing

Reply:
I was suggesting the `MuleExpressionLanguage` in particular, not the more generic `ExpressionLanguage`. At least the former already has the `isExpression` logic which checks the delimiters.

On the other hand, one downside of doing it through a non-static object is that it may be hard to get from some context without DI. For example, from the new resolvers we are creating in these PRs, we would need to provide it in the `InternalMetadataContext` too. If we wan't people to stop reimplementing the sanitize logic we should make it easy to reuse, specially considering it is just a few lines of code.

Maybe it is more convenient to just put another `ExpressionLanguageUtils` in `mule-api` and just move it there. The only problem is that it won't be able to access the [same constants](https://github.com/mulesoft/mule/blob/master/core/src/main/java/org/mule/runtime/core/api/el/ExpressionManager.java#L32) defined in `mule-core`, but we can invert that dependency.

I also just found [this other candidate](https://github.com/mulesoft/mule/blob/master/runtime-extension-model/src/main/java/org/mule/runtime/core/internal/expression/util/ExpressionUtils.java#L15). But it would be the same, we would need to repeat the impl, because there is no visibility between `runtime-extensions-model` and `mule-core`.

Reply:
let's put this in a static utility method in mule-api